### PR TITLE
Fix opencv issue #22721

### PR DESCRIPTION
**Pull request by cudawarped**
Fix https://github.com/opencv/opencv/issues/22721#issuecomment-1304837204

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/cudacodec/src/video_writer.cpp [hunk 1]
```diff
@@ -61,12 +61,6 @@ GUID EncodingProfileGuid(const EncodeProfile encodingProfile);
 GUID EncodingPresetGuid(const EncodePreset nvPreset);
 bool Equal(const GUID& g1, const GUID& g2);
 
-EncoderParams::EncoderParams() : nvPreset(ENC_PRESET_P3), tuningInfo(ENC_TUNING_INFO_HIGH_QUALITY), encodingProfile(ENC_CODEC_PROFILE_AUTOSELECT),
```
Comment (alalek) [inline modules/cudacodec/src/video_writer.cpp:64]: There are should be implementation in the `#else` branch:

```
#else // !defined HAVE_NVCUVENC

... CV_Error(...NotImplemented...) implementation stubs ...
```

Comment (asmorkalov) [reply modules/cudacodec/src/video_writer.cpp:64]: `createVideoWriter` has this else branch. With the changes parameters structure could be created correctly without linker issue, but `createVideoWriter` raises the issue. I propose to merge the PR as is.

**Review by asmorkalov — APPROVED** [review]

File: modules/cudacodec/src/video_writer.cpp (hunk 1 above)
Comment (cudawarped) [reply modules/cudacodec/src/video_writer.cpp:64]: I moved it to the header to mirror the way [cudacodec::FormatInfo()](https://github.com/opencv/opencv_contrib/blob/4.x/modules/cudacodec/include/opencv2/cudacodec.hpp#L304) is implemented.

Both structs can exist without the Nvidia Video Coded SDK being installed whether they should or not is a different matter.  Do you want me to move the implementation of both into the .cpp and prevent their use when `HAVE_NVCUVENC` and/or `HAVE_NVCUVID` are not defined?

Comment (asmorkalov) [reply modules/cudacodec/src/video_writer.cpp:64]: I prefer to have config structures always available how it's done in PR.

**Pull request merged**